tests/test_secondary.py

```diff
@@ -0,0 +1,721 @@
+      # fed in was actually for the Primary (which connects to central
+      # services) instead of for the Secondary (which obtains metadata and
+      # images via TUF from an unverified local directory, then validates
+      # them). Do this for both clients.
+      # The location of the files will be as follows, after the sample
+      # metadata archive is expanded (in test 40 below):
+
+      image_repo_mirror = ['file://' + client_dir + '/unverified/imagerepo']
+      director_mirror = ['file://' + client_dir + '/unverified/director']
+      if vin == '000':
+        # Simulate unavailable Director repo for the third Secondary
+        director_mirror[0] += '/nonexistent_directory'
+
+      repository_urls = instance.updater.pinned_metadata['repositories']
+      repository_urls['imagerepo']['mirrors'] = image_repo_mirror
+      repository_urls['director']['mirrors'] = director_mirror
+
+      # Also fix the copied pinned metadata in the individual repo updaters
+      # in the updater.
+      instance.updater.repositories['imagerepo'].mirrors = image_repo_mirror
+      instance.updater.repositories['director'].mirrors = director_mirror
+
+
+
+
+
+
+  def test_10_nonce_rotation(self):
+    """
+    Tests two uptane.clients.secondary.Secondary methods:
+      - change_nonce()
+      - set_nonce_as_sent()
+    """
+    old_nonce = secondary_instance_1.nonce_next
+
+    secondary_instance_1.change_nonce()
+    # Collision is unlikely in the next line (new random nonce equal to
+    # previous).
+    self.assertNotEqual(old_nonce, secondary_instance_1.nonce_next)
+
+
+    secondary_instance_1.set_nonce_as_sent()
+    self.assertEqual(
+        secondary_instance_1.last_nonce_sent, secondary_instance_1.nonce_next)
+
+
+
+
+
+  def test_20_validate_time_attestation(self):
+    """
+    Tests uptane.clients.secondary.Secondary::validate_time_attestation()
+    """
+
+    # Try a valid time attestation first, signed by an expected timeserver key,
+    # with an expected nonce (previously "received" from a Secondary)
+    original_time_attestation = time_attestation = {
+        'signed': {'nonces': [nonce], 'time': '2016-11-02T21:06:05Z'},
+        'signatures': [{
+          'method': 'ed25519',
+          'sig': 'aabffcebaa57f1d6397bdc5647764261fd23516d2996446c3c40b3f30efb2a4a8d80cd2c21a453e78bf99dafb9d0f5e56c4e072db365499fa5f2f304afec100e',
+          'keyid': '79c796d7e87389d1ebad04edce49faef611d139ee41ea9fb1931732afbfaac2e'}]}
+
+    # Make sure that the Secondary thinks that it sent the nonce listed in the
+    # sample data above.
+    secondary_instance_1.last_nonce_sent = nonce
+
+    if tuf.conf.METADATA_FORMAT == 'der':
+      # Convert this time attestation to the expected ASN.1/DER format.
+      time_attestation = asn1_codec.convert_signed_metadata_to_der(
+          original_time_attestation, private_key=key_timeserver_pri, resign=True)
+
+    secondary_instance_1.validate_time_attestation(time_attestation)
```

Comment:
I guess nothing happens if `time_attestation` is valid?  Might be clearer to assert that None is raised or say so in a comment.